tools/Google.Cloud.Tools.TrimCoverageReport/Program.cs

```diff
@@ -0,0 +1,124 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml.Linq;
+
+namespace TrimCoverageReport
+{
+    /// <summary>
+    /// Application to trim the detailed XML coverage report produced by dotCover to remove generated code.
+    /// </summary>
+    class Program
+    {
+        // Attributes used multiple times in the XML
+        const string TotalStatements = "TotalStatements";
+        const string CoveredStatements = "CoveredStatements";
+        const string CoveragePercent = "CoveragePercent";
+
+        static int Main(string[] args)
+        {
+            if (args.Length != 2)
+            {
+                Console.WriteLine("Arguments: <input file> <output file>");
+                return 1;
+            }
+            XDocument doc = XDocument.Load(args[0]);
+            ProcessDocument(doc);
+            // Don't use File.CreateText as that omits the byte order mark.
+            // While byte order marks are nasty, Visual Studio will add it back any time a project file is
+            // manually edited, so it's best if we follow suit.
+            using (var writer = File.CreateText(args[1]))
+            {
+                doc.Save(writer);
+            }
+            return 0;
+        }
+
+        static void ProcessDocument(XDocument document)
+        {
+            // TODO: Check how local methods are handled
+
+            // Steps:
+            // - Work out all files that are generated
+            // - Remove File elements for generated files
+            // - Remove each Statement element with a FileIndex for a generated file
+            // - Recompute the coverage for each Method element
+            // - Remove Method elements with 0 statements
+            // - Recompute the coverage for each Type element (based on descendant Method elements)
+            // - Remove Type elements with 0 statements
+            // - Recompute the coverage for each Namespace element (based on descendant Type elements)
+            // - Remove Namespace elements with 0 statements
+            // - Recompute the coverage for each Assembly element (based on descendant Namespace elements)
+            // - Remove Assembly elements with 0 statements
+
+            var fileElements = document.Root.Element("FileIndices").Elements("File");
+            var generatedFiles = fileElements.Where(file => IsGeneratedFile(file.Attribute("Name").Value)).ToList();
+            var generatedIndices = new HashSet<int>(generatedFiles.Select(file => (int) file.Attribute("Index")));
+            generatedFiles.Remove();
```

Comment:
Curious: how does this work? Doesn't `Remove` need to take semething to remove? Is there an extension method somewhere?